api/rootcerts.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+//
+// See https://github.com/hashicorp/go-rootcerts's rootcerts.go as of bb0b55efd5a1de16cb10686096b9596cfbab1fad.
+
+package api
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// CertConfig determines where LoadCACerts will load certificates from.
+type certConfig struct {
+	// CAFile is a path to a PEM-encoded certificate file or bundle.
+	CAFile string
+
+	// CACertificate is a PEM-encoded certificate or bundle.
+	CACertificate []byte
+
+	// CAPath is a path to a directory populated with PEM-encoded certificates.
+	CAPath string
+}
+
+// ConfigureTLS sets up the RootCAs on the provided tls.Config based on the CertConfig specified.
+func configureTLS(t *tls.Config, c *certConfig) error {
+	if t == nil {
+		return nil
+	}
+
+	// Greedily load all possible CA sources into the pool.
+	pool, err := loadCACerts(c)
+	if err != nil {
+		return err
+	}
+	t.RootCAs = pool
+	return nil
+}
+
+// LoadCACerts loads a CertPool from all available CA sources in CertConfig.
+func loadCACerts(c *certConfig) (*x509.CertPool, error) {
+	// Start with system CAs.
+	pool, err := loadSystemCAs()
+	if err != nil {
+		pool = x509.NewCertPool()
```

Comment:
Sorry to faff here, but I realized an implementation detail when responding to https://github.com/openbao/openbao/issues/575#issuecomment-2388451822: 

[In `x509.Certificate.Verify(...)`](https://pkg.go.dev/crypto/x509#Certificate.Verify), it says:

> If `opts.Roots` is `nil`, the platform verifier might be used, and verification details might differ from what is described below. If system roots are unavailable the returned error will be of type `SystemRootsError`. 

So we probably want to verify if we've _added_ anything successfully to the pool, and if so, return `pool`. But if we didn't have anything to add to the pool, and we fell through to this line because the system pool failed... We'd now be returning a valid-but-empty pool on line 75 which means publicly trusted certs would fail to verify.

(If, for instance, there was some issue with System CAs from our call but the next call in Verify worked. idk.)

I think it'd be good to avoid that scenario and let `pool` be `nil` in that case... That does mean keeping track of an `added` bool or something and making all of the `if err := append....` calls into a `added, err := ...` format.

Does that sound right? Now I'm not sure if I like the return-the-system-pool-by-default behavior after all, given that it is implied... Hm! 

Thoughts?

Reply:
I like that idea!